src/lib.rs

```diff
@@ -0,0 +1,135 @@
+/// in a browser
+pub fn process(descriptor: String, network: Network) -> Result<String> {
+    let descriptor_string: Descriptor<String> = descriptor.parse()?;
+    debug!("descriptor_string: {}", descriptor_string);
+
+    let (keys, address) = create_key_pairs_and_address(&descriptor_string, network)?;
+
+    let wallet_data = create_wallet_data(address, keys, &descriptor_string)?;
+
+    let html = paper_wallets(&wallet_data)?;
+
+    Ok(to_data_url(&html, "text/html"))
+}
+
+/// Creates a key pair for every alias in the given descriptor,
+/// Creates another descriptor replacing alias with the relative public key, so that we can compute the address.
+pub fn create_key_pairs_and_address(
+    descriptor_string: &Descriptor<String>,
+    network: Network,
+) -> Result<(HashMap<String, WifAndHexPub>, Address)> {
+    let mut rng = thread_rng();
+    let secp = secp256k1::Secp256k1::signing_only();
+    let mut keys = HashMap::new();
+    let alias_to_key = |alias: &String| -> Result<PublicKey> {
+        let entropy: [u8; 32] = rng.gen();
+        let sk = PrivateKey {
+            compressed: true,
+            network,
+            key: SecretKey::from_slice(&entropy)?,
+        };
+        let key = PublicKey::from_private_key(&secp, &sk);
+        keys.insert(
+            alias.clone(),
+            WifAndHexPub {
+                wif: sk.to_wif(),
+                hex_pub: key.to_string(),
+            },
+        );
+
+        Ok(key)
+    };
+    let descriptor_keys: Descriptor<PublicKey> =
+        descriptor_string.translate_pk(alias_to_key, |_| unreachable!())?;
+    debug!("descriptor_keys: {}", descriptor_keys);
+    debug!("key_map: {:?}", keys);
+    let address = descriptor_keys.address(network)?;
+    debug!("address: {}", address.to_string());
+    Ok((keys, address))
+}
+
+/// Creates data for every single paper wallet (which is different according to the relative owner)
+fn create_wallet_data(
+    address: Address,
+    keys: HashMap<String, WifAndHexPub>,
+    descriptor_alias: &Descriptor<String>,
+) -> Result<Vec<WalletData>> {
+    let address_string = address.to_string();
+    let address_qr = match address.address_type() {
+        Some(AddressType::P2wpkh) | Some(AddressType::P2wsh) => address_string.to_ascii_uppercase(),
+        _ => address_string.clone(),
+    };
+    let descriptor_alias_string = descriptor_alias
+        .to_string()
+        .split('#')
+        .next()
+        .map(|s| s.to_string())
+        .unwrap();
+
+    let mut results = vec![];
+    for alias in keys.keys() {
+        let mut legend = HashMap::new();
+        for (alias_internal, wif_and_pub) in keys.iter() {
+            debug!("int:{}", alias_internal);
+            if alias == alias_internal {
+                legend.insert(alias_internal, wif_and_pub.wif.clone());
+            } else {
+                legend.insert(alias_internal, wif_and_pub.hex_pub.clone());
+            }
+        }
```

Comment:
i generally prefer to use iterators whenever i can instead of making variables mutable.. but that's more of my personal preference, feel free to ignore this if you don't like it as much.

```rust
keys
  .iter()
  .map(|(alias_internal, wif_and_pub)| if alias == alias_internal {
      (alias_internal, wif_and_pub.wif.clone())
    } else {
      (alias_internal, wif_and_pub.hex_pub.clone())
    }
  )
  .collect::<HashMap<_, _>>();
```

Reply:
changed to avoid mut

I also prefer that but I noticed many times when I write some algo for the first time, l use explicit for loop and then refactor with iterators chain, maybe less cognitive effort for me.